src/Entity/BannerSelection/Campaign.php

```diff
@@ -38,12 +53,16 @@ public function isInActiveDateRange( \DateTime $time ): bool {
 			$time->getTimestamp() <= $this->end->getTimestamp();
 	}
 
-	public function selectBucket( ?string $bucketId, callable $fallbackSelectionStrategy ): Bucket {
+	public function selectBucket( ?string $bucketId ): Bucket {
 		foreach ( $this->buckets as $bucket ) {
 			if ( $bucket->getIdentifier() === $bucketId ) {
 				return $bucket;
 			}
 		}
-		return $fallbackSelectionStrategy( $this->buckets );
+		return $this->buckets[$this->rng->getRandomInteger( 0, count( $this->buckets ) - 1 )];
+	}
+
+	public function isRatioLimited(): bool {
```

Comment:
It feels slightly weird to have a name that hints at a static, predictable getter and the internal logic being random. But my tired brain can't come up with a better suggestion, so we can leave it for now.